139 dont mark sms permanently failed

latin-panda opened the pull request:
# Description

This PR will introduce a soft fail with retry process for SMS that works as follows:

1.. If It’s temporary error, soft fail the sms, this means:
        1.1 SMS will be back to `UNSENT` status, so Gateway will pick it up again automatically
        1.2 SMS’ retries counter increases by 1
        1.3 Log that X error occurred and that will retry for the X time in X minutes

2.. If sms’ maximum of retries are reached then
        2.1 Hard fail: SMS status will be `FAILED` and stop retries
        2.2 Log error
        2.3 Counter retries reset to 0

3.. At this point the user can manually select the message and press `Retry` button, if necessary
        3.1 If they do, then it goes back to step # 1 in case the SMS fails again…

I couldn't add any [new error code](https://developer.android.com/reference/android/telephony/SmsManager#RESULT_RIL_NETWORK_ERR) for SMS, since we are supporting older versions of Android and most of the nice ones were added in SDK 30.

**This PR is dependant of this one:**  https://github.com/medic/medic-gateway/pull/158

**To AT:** My test device at the moment is Android 9. Can you please test this in older Android version? Thanks!

https://github.com/medic/medic-gateway/issues/139

# Code review items

- Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- Documented: Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

latin-panda commented:
Hi @dianabarsan , could you please review this code? :)

dianabarsan commented on src/test/java/medic/gateway/alert/SmsSenderTest.java:
```diff
@@ -70,8 +70,8 @@ public void sendUnsentSmses_shouldPassUnsentMessagesToSmsManager() {
-					cols("_id",     "status", "last_action", "_to", "content"),
-					vals("id-" + s, s,        0,             "+1",  "testing: " + s));
+					cols("_id",     "status", "last_action", "_to", "content", "retries"),
+					vals("id-" + s, s,        0,             "+1",  "testing: " + s, 0));
```
Is it possible to add some tests here to check if the retry mechanism is successful:
- skip retry when it's not time
- do retry when it's time
- retry interval calculation for multiple retries
- no retry when exceeding max nbr.

latin-panda replied:
Thanks! Done

latin-panda commented on README.md:
```diff
@@ -191,6 +191,21 @@ The `message` property may be logged and/or displayed to users in the `medic-gat
+1. A possible temporary error occurs and Gateway retries sending the SMS:
+    1.1 SMS status will be updated to `UNSENT`, so Gateway will find it and add it into the `send queue` automatically.
+    1.2 SMS' `retry counter` increases by 1.
+    1.3 Gateway logs that X error occurred and that will retry for the X time in X minutes.
```
Done!

latin-panda commented on README.md:
```diff
@@ -191,6 +191,21 @@ The `message` property may be logged and/or displayed to users in the `medic-gat
+    1.2 SMS' `retry counter` increases by 1.
+    1.3 Gateway logs that X error occurred and that will retry for the X time in X minutes.
+
+2. If the SMS retry limit is reached, then:
```
Done!

latin-panda commented on src/androidTest/java/medic/gateway/alert/IntentProcessorInstrumentationTest.java:
```diff
@@ -154,17 +196,38 @@ public void test_onReceive_NULL_PDU_shouldUdpateSendStatusAndIncludeErrorCodeInR
+	public void test_onReceive_NULL_PDU_shouldUdpateSendStatusAndRetryAfterSoftFail() throws Exception {
+		// given
+		String id = randomUuid();
+		aWoMessageIsInDbWith(id, PENDING, 0);
```
Thanks! Done

latin-panda commented:
Hi @garethbowen,  regarding your comments in https://github.com/medic/medic-gateway/pull/161, I have done the changes in this branch since it's about the retry process:
1. Added the migration and updated the upgrade method from DB.
2. Added the power of 1.5 to the calculation of retry waiting time.
